velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -202,310 +195,218 @@ inline static TResult addLarge(
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        bool overflow = false;
-        Operation::template apply<R, A, B>(
-            rawResults[row],
-            constant,
-            rawValues[row],
-            aRescale_,
-            bRescale_,
-            aPrecision_,
-            aScale_,
-            bPrecision_,
-            bScale_,
-            rPrecision_,
-            rScale_,
-            overflow);
-        if (overflow ||
-            !velox::DecimalUtil::valueInPrecisionRange(
-                rawResults[row], rPrecision_)) {
-          result->setNull(row, true);
-        }
-      });
-    } else if (args[0]->isFlatEncoding() && args[1]->isConstantEncoding()) {
-      // Fast path for (flat, const).
-      auto flatValues = args[0]->asUnchecked<FlatVector<A>>();
-      auto constant = args[1]->asUnchecked<SimpleVector<B>>()->valueAt(0);
-      auto rawValues = flatValues->mutableRawValues();
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        bool overflow = false;
-        Operation::template apply<R, A, B>(
-            rawResults[row],
-            rawValues[row],
-            constant,
-            aRescale_,
-            bRescale_,
-            aPrecision_,
-            aScale_,
-            bPrecision_,
-            bScale_,
-            rPrecision_,
-            rScale_,
-            overflow);
-        if (overflow ||
-            !velox::DecimalUtil::valueInPrecisionRange(
-                rawResults[row], rPrecision_)) {
-          result->setNull(row, true);
-        }
-      });
-    } else if (args[0]->isFlatEncoding() && args[1]->isFlatEncoding()) {
-      // Fast path for (flat, flat).
-      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
-      auto rawA = flatA->mutableRawValues();
-      auto flatB = args[1]->asUnchecked<FlatVector<B>>();
-      auto rawB = flatB->mutableRawValues();
-
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        bool overflow = false;
-        Operation::template apply<R, A, B>(
-            rawResults[row],
-            rawA[row],
-            rawB[row],
-            aRescale_,
-            bRescale_,
-            aPrecision_,
-            aScale_,
-            bPrecision_,
-            bScale_,
-            rPrecision_,
-            rScale_,
-            overflow);
-        if (overflow ||
-            !velox::DecimalUtil::valueInPrecisionRange(
-                rawResults[row], rPrecision_)) {
-          result->setNull(row, true);
-        }
-      });
+// Add the value a and b, set the result r.
+// Suppose aScale is more than bScale, then aRescale is 0, bRescale is `aScale -
+// bScale`, when the result precision and scale do not adjust(result precision
+// is less than LongDecimalType::kMaxPrecision), the result scale is
+// max(aScale, bScale), so we rescale the value with small scale to the result
+// scale.
+// The overflow flag will be set to true if an overflow occurs during the
+// addition.
```

Comment:
Comment suggestion:

> // Adds the values 'a' and 'b' and stores the result in 'r'. To align the scales
// of inputs, the value with the smaller scale is rescaled to the larger scale.
// 'aRescale' and 'bRescale' are the rescale factors needed to rescale 'a' and
// 'b'. 'rPrecision' and 'rScale' are the precision and scale of the result.
// The overflow flag is set to true if an overflow occurs during the addition.